File: StriverSDE/Arrays1/set-matrix-zero/set-matrix-zero.go
// Problem Statement: Given a matrix if an element in the matrix is 0 then you will have to set its entire column and row to 0 and then return the matrix.
// Examples 1:
// Input: matrix=[[1,1,1],[1,0,1],[1,1,1]]
// Output: [[1,0,1],[0,0,0],[1,0,1]]
// Explanation: Since matrix[2][2]=0.Therefore the 2nd column and 2nd row wil be set to 0.
// Input: matrix=[[0,1,2,0],[3,4,5,2],[1,3,1,5]]
// Output:[[0,0,0,0],[0,4,5,0],[0,3,1,0]]
// Explanation:Since matrix[0][0]=0 and matrix[0][3]=0. Therefore 1st row, 1st column and 4th column will be set to 0
package main

import "fmt"

func main() {
	// declare a 2d array
	matrix1 := [][]int{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}
	matrix2 := [][]int{{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}}
	// call the function
	fmt.Println(setZeroesBruteForce(matrix1))
	fmt.Println(setZeroesBruteForce(matrix2))
	// fmt.Println(setZeroes(matrix2))
	// fmt.Println(setZeroes(matrix2))

}

// setZeroesBruteForce function
func setZeroesBruteForce(nums [][]int) [][]int {
	mapRow := make(map[int]bool)
	mapCol := make(map[int]bool)
	for i := 0; i < len(nums); i++ {
		for j := 0; j < len(nums[i]); j++ {
			if nums[i][j] == 0 {
				mapRow[i] = true
				mapCol[j] = true
			}
		}
	}
	for i := 0; i < len(nums); i++ {
		for j := 0; j < len(nums[i]); j++ {
			if mapRow[i] || mapCol[j] {
				nums[i][j] = 0
			}
		}
	}
	return nums
}

File: striver/set-matrix-zero.ts
// Problem Statement: Given a matrix if an element in the matrix is 0 then you will have to set its entire column and row to 0 and then return the matrix.

const setMatrixZero = (matrix: number[][]) => {
  let firstRowHasZero = false;
  let firstColHasZero = false;

  for (let i = 0; i < matrix.length; i++) {
    if (matrix[i][0] === 0) firstColHasZero = true;
  }
  for (let j = 0; j < matrix[0].length; j++) {
    if (matrix[0][j] === 0) firstRowHasZero = true;
  }
  for (let i = 1; i < matrix.length; i++) {
    for (let j = 1; j < matrix[0].length; j++) {
      if (matrix[i][j] === 0) {
        matrix[i][0] = 0;
        matrix[0][j] = 0;
      }
    }
  }
  for (let i = 1; i < matrix.length; i++) {
    if (matrix[i][0] === 0)
      for (let j = 1; j < matrix[0].length; j++) {
        matrix[i][j] = 0;
      }
  }
  for (let j = 1; j < matrix[0].length; j++) {
    if (matrix[0][j] === 0)
      for (let i = 1; i < matrix.length; i++) {
        matrix[i][j] = 0;
      }
  }
  if (firstRowHasZero)
    for (let j = 0; j < matrix[0].length; j++) {
      console.log(0, j, matrix[0][j]);
      matrix[0][j] = 0;
    }
  if (firstColHasZero)
    for (let i = 0; i < matrix.length; i++) {
      matrix[i][0] = 0;
    }

  return matrix;
};
console.log(
  setMatrixZero([
    [0, 1, 2, 0],
    [3, 4, 5, 2],
    [1, 3, 1, 5],
  ])
); //[[0,0,0,0],[0,4,5,0],[0,3,1,0]]

// console.log(
//   setMatrixZero([
//     [1, 1, 1],
//     [1, 0, 1],
//     [1, 1, 1],
//   ])
// ); // [[1,0,1],[0,0,0],[1,0,1]]
